### PR TITLE
Proposal to implement GetAuthenticationResultForUserAsync for EasyAuth

### PR DESCRIPTION
**Pull request by jmprieur**
implement GetAuthenticationResultForUserAsync for EasyAuth

Still need to:
- process the static analysis issues
- test with easy auth

### CODE REVIEW

File: src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationHandler.cs [hunk 1]
```diff
@@ -35,28 +35,15 @@ public class AppServicesAuthenticationHandler : AuthenticationHandler<AppService
...
         /// <inheritdoc/>
         protected override Task<AuthenticateResult> HandleAuthenticateAsync()
         {
             if (AppServicesAuthenticationInformation.IsAppServicesAadAuthenticationEnabled)
             {
-                string? idToken = GetIdToken();
-                string? idp = GetIdp();
-
-                if (idToken != null && idp != null)
+                ClaimsPrincipal claimsPrincipal = AppServicesAuthenticationInformation.GetUser(Context.Request.Headers);
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationHandler.cs:44]: Most was moved to AppServicesAuthenticationInformation as it's also used in AppServicesAuthenticationTokenAcquisition now.

File: src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationTokenAcquisition.cs [hunk 2]
```diff
@@ -126,7 +126,22 @@ private string GetAccessToken(IHeaderDictionary? headers)
...
-            throw new NotImplementedException();
+            string? idToken = AppServicesAuthenticationInformation.GetIdToken(CurrentHttpContext?.Request?.Headers);
+            ClaimsPrincipal? userClaims = AppServicesAuthenticationInformation.GetUser(CurrentHttpContext?.Request?.Headers);
+            string accessToken = await GetAccessTokenForUserAsync(scopes, tenantId, userFlow, user, tokenAcquisitionOptions);
+
+            AuthenticationResult authenticationResult = new AuthenticationResult(
+                accessToken,
+                isExtendedLifeTimeToken: false,
+                userClaims?.GetDisplayName(),
+                DateTimeOffset.Now,
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationTokenAcquisition.cs:137]: We need to check if we can do better for the times.

File: src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationTokenAcquisition.cs [hunk 3]
```diff
@@ -126,7 +126,22 @@ private string GetAccessToken(IHeaderDictionary? headers)
...
+                DateTimeOffset.Now,
+                userClaims?.GetTenantId(),
+                null,
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationTokenAcquisition.cs:140]: We can check if we could instantiate an IAccount

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 4]
```diff
@@ -304,6 +304,10 @@ internal class TokenAcquisition : ITokenAcquisitionInternal
                 builder.WithCorrelationId(tokenAcquisitionOptions.CorrelationId);
                 builder.WithForceRefresh(tokenAcquisitionOptions.ForceRefresh);
                 builder.WithClaims(tokenAcquisitionOptions.Claims);
+                if (tokenAcquisitionOptions.PoPConfiguration != null)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:307]: I think that this was missing

File: src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationInformation.cs [hunk 5]
```diff
@@ -91,5 +97,87 @@ public static bool IsAppServicesAadAuthenticationEnabled
             return headerPlusValue?.Substring(header.Length + 1);
         }
 #endif
+
+        /// <summary>
+        /// Get the ID token from the headers send by App services authentication.
+        /// </summary>
+        /// <param name="headers">Headers.</param>
+        /// <returns>the ID Token.</returns>
+        public static string? GetIdToken(IDictionary<string, StringValues> headers)
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationInformation.cs:106]: make private

File: src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationInformation.cs [hunk 6]
```diff
@@ -91,5 +97,87 @@ public static bool IsAppServicesAadAuthenticationEnabled
...
+#endif
+            return idToken;
+        }
+
+          /// <summary>
+        /// Get the IDP token from the headers send by App services authentication.
+        /// </summary>
+        /// <param name="headers">Headers.</param>
+        /// <returns>the IDP.</returns>
+        public static string? GetIdp(IDictionary<string, StringValues> headers)
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationInformation.cs:132]: make private

File: src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationInformation.cs [hunk 7]
```diff
@@ -91,5 +97,87 @@ public static bool IsAppServicesAadAuthenticationEnabled
...
+#endif
+            return idp;
+        }
+
+        /// <summary>
+        /// Get the user claims from the headers and environment variables
+        /// </summary>
+        /// <param name="headers">Headers</param>
+        /// <returns>User claims</returns>
+        public static ClaimsPrincipal? GetUser(IDictionary<string, StringValues> headers)
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/AppServicesAuth/AppServicesAuthenticationInformation.cs:158]: make private

File: src/Microsoft.Identity.Web/TokenAcquisition.cs (hunk 4 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenAcquisition.cs:307]: hmmm...

**Pull request merged**